com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java

```diff
@@ -0,0 +1,328 @@
+import com.microsoft.java.debug.core.IDebugSession;
+import com.microsoft.java.debug.core.adapter.IDebugAdapterContext;
+import com.microsoft.java.debug.core.adapter.IHotCodeReplaceListener;
+import com.microsoft.java.debug.core.adapter.IHotCodeReplaceProvider;
+import com.microsoft.java.debug.core.protocol.Events.DebugEvent;
+
+/**
+ * The hot code replace provider listens for changes to class files and notifies
+ * the running debug session of the changes.
+ */
+public class JavaHotCodeReplaceProvider implements IHotCodeReplaceProvider, IResourceChangeListener {
+
+    private static final Logger logger = Logger.getLogger(Configuration.LOGGER_NAME);
+
+    private static final String CLASS_FILE_EXTENSION = "class"; //$NON-NLS-1$
+
+    private IDebugSession currentDebugSession;
+
+    private IDebugAdapterContext context;
+
+    private List<IHotCodeReplaceListener> listeners = new ArrayList<>();
+
+    /**
+     * Visitor for resource deltas.
+     */
+    protected ChangedClassFilesVisitor classFilesVisitor = new ChangedClassFilesVisitor();
+
+    /**
+     * A visitor which collects changed class files.
+     */
+    class ChangedClassFilesVisitor implements IResourceDeltaVisitor {
+        /**
+         * The collection of changed class files.
+         */
+        protected List<IResource> changedFiles = null;
+
+        /**
+         * Collection of qualified type names, corresponding to class files.
+         */
+        protected List<String> fullyQualifiedNames = null;
+
+        /**
+         * Answers whether children should be visited.
+         * <p>
+         * If the associated resource is a class file which has been changed, record it.
+         * </p>
+         */
+        @Override
+        public boolean visit(IResourceDelta delta) {
+            if (delta == null || 0 == (delta.getKind() & IResourceDelta.CHANGED)) {
+                return false;
+            }
+            IResource resource = delta.getResource();
+            if (resource != null) {
+                switch (resource.getType()) {
+                    case IResource.FILE:
+                        if (0 == (delta.getFlags() & IResourceDelta.CONTENT)) {
+                            return false;
+                        }
+                        if (CLASS_FILE_EXTENSION.equals(resource.getFullPath().getFileExtension())) {
+                            IPath localLocation = resource.getLocation();
+                            if (localLocation != null) {
+                                String path = localLocation.toOSString();
+                                IClassFileReader reader = ToolFactory.createDefaultClassFileReader(path,
+                                        IClassFileReader.CLASSFILE_ATTRIBUTES);
+                                if (reader != null) {
+                                    // this name is slash-delimited
+                                    String qualifiedName = new String(reader.getClassName());
```

Comment:
why new a string instance? string is immutable

Reply:
I think we should reuse the source name info from .class file since fqn is not mapped one-to-one to the java file.

Reply:
reader.getClassName is a char[]